stats/statsaxis.cpp

```diff
@@ -0,0 +1,394 @@
+// margins.
+int StatsAxis::guessNumTicks(const QtCharts::QChart *chart, const QtCharts::QAbstractAxis *axis, const std::vector<QString> &strings) const
+{
+	QFont font = axis->labelsFont();
+	QFontMetrics fm(font);
+	int minSize = fm.height();
+	for (const QString &s: strings) {
+		QSize size = fm.size(Qt::TextSingleLine, s);
+		int needed = horizontal ? size.width() : size.height();
+		if (needed > minSize)
+			minSize = needed;
+	}
+
+	// Add space between labels
+	if (horizontal)
+		minSize = minSize * 3 / 2;
+	else
+		minSize *= 2;
+	QRectF chartSize = chart->plotArea();
+	double availableSpace = horizontal ? chartSize.width() : chartSize.height();
+	int numTicks = lrint(availableSpace / minSize);
+	return std::max(numTicks, 2);
+}
+
+ValueAxis::ValueAxis(double min, double max, int decimals, bool horizontal) : StatsAxisTemplate(horizontal),
+	min(min), max(max), decimals(decimals)
+{
+}
+
+std::pair<double, double> ValueAxis::minMax() const
+{
+	return { QValueAxis::min(), QValueAxis::max() };
+}
+
+static QString makeFormatString(int decimals)
+{
+	return QStringLiteral("%.%1f").arg(decimals < 0 ? 0 : decimals);
+}
+
+void ValueAxis::updateLabels(const QtCharts::QChart *chart)
+{
+	using QtCharts::QValueAxis;
+
+	// Avoid degenerate cases
+	if (max - min < 0.0001) {
+		max = 0.5;
+		min = -0.5;
+	}
+
+	QLocale loc;
+	QString minString = loc.toString(min, 'f', decimals);
+	QString maxString = loc.toString(max, 'f', decimals);
+	int numTicks = guessNumTicks(chart, this, { minString, maxString});
+
+	// Use full decimal increments
+	double height = max - min;
+	double inc = height / numTicks;
+	double digits = floor(log10(inc));
+	int digits_int = lrint(digits);
+	double digits_factor = pow(10.0, digits);
+	int inc_int = std::max((int)ceil(inc / digits_factor), 1);
+	// Do "nice" increments: 1, 2, 4, 5.
+	if (inc_int > 5)
+		inc_int = 10;
+	if (inc_int == 3)
+		inc_int = 5;
+	inc = inc_int * digits_factor;
+	if (-digits_int > decimals)
+		decimals = -digits_int;
+
+	setLabelFormat(makeFormatString(decimals));
+	double actMin = floor(min /  inc) * inc;
+	double actMax = ceil(max /  inc) * inc;
+	int num = lrint((actMax - actMin) / inc);
+	setRange(actMin, actMax);
+	setTickCount(num + 1);
+}
+
+CountAxis::CountAxis(int count, bool horizontal) : ValueAxis(0.0, (double)count, 0, horizontal),
+	count(count)
+{
+}
+
+void CountAxis::updateLabels(const QtCharts::QChart *chart)
+{
+	QLocale loc;
+	QString countString = loc.toString(count);
+	int numTicks = guessNumTicks(chart, this, { countString });
+
+	// Get estimate of step size
+	if (count <= 0)
+		count = 1;
+	// When determining the step size, make sure to round up
+	int step = (count + numTicks - 1) / numTicks;
+	if (step <= 0)
+		step = 1;
+
+	// Get the significant first or first two digits
+	int scale = 1;
+	int significant = step;
+	while (significant > 25) {
+		significant /= 10;
+		scale *= 10;
+	}
+
+	for (int increment: { 1, 2, 4, 5, 10, 15, 20, 25 }) {
```

Comment:
I'm not sure that 4 is a very common increment... 0, 4, 8, 12 , 16... ok, maybe it's not bad...